File: fsd_ws/src/fsd_stack/fsd_stack/safety_supervisor_node.py
"""BLOCK 8 — SAFETY SUPERVISOR (owner: Controls & Actuation).

Watches /safety/heartbeat from every required node plus plausibility checks
on the data streams. Any violation latches /safety/ebs_trigger = True.
Publishes a 10 Hz keepalive (False) when healthy — the STM32 treats a
missing keepalive as a trigger condition, so a dead supervisor is itself
an EBS trigger.

This is the SOFTWARE stop path (path 3 of 3). It must never be the only
one: the RES/SDC hardware path and the STM32 command-timeout firmware path
exist independently of everything in this file.
"""

import math

import rclpy
from rclpy.node import Node

from std_msgs.msg import Bool
from nav_msgs.msg import Odometry
from fsd_msgs.msg import Heartbeat, ConeMap, VehicleCmd

from .common import HeartbeatEmitter, qos_reliable

HEARTBEAT_TIMEOUT_S = 0.5
BLIND_SPEED_MPS = 0.5
BLIND_GRACE_S = 3.0
MAX_POSE_JUMP_M = 1.0


class SafetySupervisorNode(Node):
    def __init__(self):
        super().__init__('safety_supervisor')
        # Which nodes MUST be alive. Set per launch config: the sim launch
        # excludes perception nodes because the sim publishes ideal cones.
        self.declare_parameter('required_nodes', [
            'cone_detection', 'cone_localization', 'state_estimation',
            'cone_mapping', 'path_planning', 'motion_control'])
        self.declare_parameter('max_steering_rad', 0.35)

        self._required = list(self.get_parameter('required_nodes').value)
        self._steer_max = float(self.get_parameter('max_steering_rad').value)

        self._last_hb = {}          # node_id -> (t, status)
        self._triggered = False
        self._start_t = self.get_clock().now().nanoseconds * 1e-9
        self._last_pose = None
        self._speed = 0.0
        self._cone_count = 0
        self._blind_since = None

        self.create_subscription(Heartbeat, '/safety/heartbeat',
                                 self._on_hb, qos_reliable(10))
        self.create_subscription(Odometry, '/odometry/filtered',
                                 self._on_odom, qos_reliable(10))
        self.create_subscription(ConeMap, '/mapping/track',
                                 self._on_map, qos_reliable(5))
        self.create_subscription(VehicleCmd, '/control/cmd',
                                 self._on_cmd, qos_reliable(1))

        self._pub = self.create_publisher(Bool, '/safety/ebs_trigger',
                                          qos_reliable(10))
        self.create_timer(0.05, self._check)      # 20 Hz checks
        self.create_timer(0.1, self._keepalive)   # 10 Hz keepalive
        self._hb = HeartbeatEmitter(self, 'safety_supervisor')

    # ------------------------------------------------------------- inputs
    def _now(self):
        return self.get_clock().now().nanoseconds * 1e-9

    def _on_hb(self, msg: Heartbeat):
        self._last_hb[msg.node_id] = (self._now(), msg.status)
        if msg.status == Heartbeat.STATUS_ERROR:
            self._trigger(f'node {msg.node_id} reports ERROR: {msg.message}')

    def _on_odom(self, msg: Odometry):
        p = msg.pose.pose.position
        v = msg.twist.twist.linear.x
        if any(not math.isfinite(f) for f in (p.x, p.y, v)):
            self._trigger('NaN/inf in odometry')
            return
        if self._last_pose is not None:
            jump = math.hypot(p.x - self._last_pose[0], p.y - self._last_pose[1])
            if jump > MAX_POSE_JUMP_M:
                self._trigger(f'pose jump {jump:.2f} m between odom updates')
        self._last_pose = (p.x, p.y)
        self._speed = v

    def _on_map(self, msg: ConeMap):
        self._cone_count = len(msg.cones)

    def _on_cmd(self, msg: VehicleCmd):
        if any(not math.isfinite(f) for f in
               (msg.steering_angle, msg.torque_request, msg.brake_cmd)):
            self._trigger('NaN/inf in control command')
            return
        if abs(msg.steering_angle) > self._steer_max * 1.05:
            self._trigger(f'steering command {msg.steering_angle:.2f} rad '
                          f'exceeds physical limit')

    # ------------------------------------------------------------- checks
    def _check(self):
        now = self._now()
        # Give nodes a startup grace period before demanding heartbeats.
        if now - self._start_t > 5.0:
            for node_id in self._required:
                seen = self._last_hb.get(node_id)
                if seen is None:
                    self._trigger(f'node {node_id} never sent a heartbeat')
                elif now - seen[0] > HEARTBEAT_TIMEOUT_S:
                    self._trigger(f'node {node_id} heartbeat missing '
                                  f'{now - seen[0]:.2f} s')

        # Moving blind: speed > threshold with zero confirmed cones.
        if self._speed > BLIND_SPEED_MPS and self._cone_count == 0:
            if self._blind_since is None:
                self._blind_since = now
            elif now - self._blind_since > BLIND_GRACE_S:
                self._trigger('moving with zero confirmed cones for > 3 s')
        else:
            self._blind_since = None

    def _keepalive(self):
        msg = Bool()
        msg.data = self._triggered
        self._pub.publish(msg)

    def _trigger(self, reason):
        if not self._triggered:
            self._triggered = True   # latched until process restart, per spec
            self.get_logger().error(f'EBS TRIGGERED: {reason}')
            self._hb.set_status(Heartbeat.STATUS_ERROR, f'EBS: {reason}')
            msg = Bool()
            msg.data = True
            self._pub.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = SafetySupervisorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: fsd_ws/src/fsd_cpp/src/safety_supervisor_node.cpp
// BLOCK 8 — SAFETY SUPERVISOR (C++). Heartbeat watchdog + plausibility
// checks; latched /safety/ebs_trigger with 10 Hz keepalive (false when
// healthy) so a dead supervisor is itself a trigger condition downstream.

#include <cmath>
#include <map>
#include <memory>
#include <string>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/bool.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <fsd_msgs/msg/heartbeat.hpp>
#include <fsd_msgs/msg/cone_map.hpp>
#include <fsd_msgs/msg/vehicle_cmd.hpp>

#include "fsd_cpp/common.hpp"

using fsd_msgs::msg::Heartbeat;

class SafetySupervisorNode : public rclcpp::Node
{
public:
  SafetySupervisorNode() : Node("safety_supervisor")
  {
    declare_parameter<std::vector<std::string>>("required_nodes", {
      "stereo_cone", "state_estimation", "cone_mapping",
      "path_planning", "motion_control"});
    declare_parameter<double>("max_steering_rad", 0.35);
    // How long the car may move with an empty map before this is an emergency.
    // This covers driving blind, but it must not fire during BOOTSTRAP: the
    // mapper needs N_CONFIRM sightings of a cone before it publishes anything,
    // so a car that has just started creeping legitimately has an empty map for
    // a few seconds. At 3 s it was killing runs at t=6 s, before the first cone
    // was ever confirmed. 8 s is ~20 m at exploration speed and still well
    // inside the 30 s standstill limit the rules impose (FB2027 D2.6.1).
    declare_parameter<double>("blind_timeout_s", 8.0);
    required_ = get_parameter("required_nodes").as_string_array();
    steer_max_ = get_parameter("max_steering_rad").as_double();
    blind_timeout_ = get_parameter("blind_timeout_s").as_double();
    start_t_ = now().seconds();

    hb_sub_ = create_subscription<Heartbeat>(
      "/safety/heartbeat", fsd::qos_reliable(10),
      [this](Heartbeat::ConstSharedPtr m) {
        last_hb_[m->node_id] = now().seconds();
        if (m->status == Heartbeat::STATUS_ERROR) {
          trigger("node " + m->node_id + " reports ERROR: " + m->message);
        }
      });
    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      "/odometry/filtered", fsd::qos_reliable(10),
      [this](nav_msgs::msg::Odometry::ConstSharedPtr m) { on_odom(m); });
    map_sub_ = create_subscription<fsd_msgs::msg::ConeMap>(
      "/mapping/track", fsd::qos_reliable(5),
      [this](fsd_msgs::msg::ConeMap::ConstSharedPtr m) {
        cone_count_ = m->cones.size();
      });
    cmd_sub_ = create_subscription<fsd_msgs::msg::VehicleCmd>(
      "/control/cmd", fsd::qos_reliable(1),
      [this](fsd_msgs::msg::VehicleCmd::ConstSharedPtr m) {
        if (!std::isfinite(m->steering_angle) ||
            !std::isfinite(m->torque_request) || !std::isfinite(m->brake_cmd))
        {
          trigger("NaN/inf in control command");
        } else if (std::abs(m->steering_angle) > steer_max_ * 1.05) {
          trigger("steering command exceeds physical limit");
        }
      });

    pub_ = create_publisher<std_msgs::msg::Bool>(
      "/safety/ebs_trigger", fsd::qos_reliable(10));
    check_timer_ = create_wall_timer(std::chrono::milliseconds(50),
                                     [this]() { check(); });
    keepalive_timer_ = create_wall_timer(std::chrono::milliseconds(100),
                                         [this]() { keepalive(); });
    hb_ = std::make_unique<fsd::HeartbeatEmitter>(this, "safety_supervisor");
  }

private:
  void on_odom(nav_msgs::msg::Odometry::ConstSharedPtr m)
  {
    const double x = m->pose.pose.position.x;
    const double y = m->pose.pose.position.y;
    const double v = m->twist.twist.linear.x;
    if (!std::isfinite(x) || !std::isfinite(y) || !std::isfinite(v)) {
      trigger("NaN/inf in odometry");
      return;
    }
    if (have_pose_) {
      const double jump = std::hypot(x - last_x_, y - last_y_);
      if (jump > 1.0) {
        trigger("pose jump " + std::to_string(jump) + " m between odom updates");
      }
    }
    last_x_ = x;
    last_y_ = y;
    have_pose_ = true;
    speed_ = v;
  }

  void check()
  {
    const double t = now().seconds();
    if (t - start_t_ > 5.0) {   // startup grace
      for (const auto & id : required_) {
        const auto it = last_hb_.find(id);
        if (it == last_hb_.end()) {
          trigger("node " + id + " never sent a heartbeat");
        } else if (t - it->second > 0.5) {
          trigger("node " + id + " heartbeat missing");
        }
      }
    }
    if (speed_ > 0.5 && cone_count_ == 0) {
      if (blind_since_ < 0.0) {
        blind_since_ = t;
      } else if (t - blind_since_ > blind_timeout_) {
        trigger("moving with zero confirmed cones for > " +
                std::to_string(static_cast<int>(blind_timeout_)) + " s");
      }
    } else {
      blind_since_ = -1.0;
    }
  }

  void keepalive()
  {
    std_msgs::msg::Bool msg;
    msg.data = triggered_;
    pub_->publish(msg);
  }

  void trigger(const std::string & reason)
  {
    if (triggered_) {
      return;
    }
    triggered_ = true;   // latched until restart
    RCLCPP_ERROR(get_logger(), "EBS TRIGGERED: %s", reason.c_str());
    hb_->set_status(Heartbeat::STATUS_ERROR, "EBS: " + reason);
    std_msgs::msg::Bool msg;
    msg.data = true;
    pub_->publish(msg);
  }

  std::vector<std::string> required_;
  double steer_max_, start_t_;
  std::map<std::string, double> last_hb_;
  bool triggered_{false}, have_pose_{false};
  double last_x_{0.0}, last_y_{0.0}, speed_{0.0}, blind_since_{-1.0};
  double blind_timeout_{8.0};
  size_t cone_count_{0};

  rclcpp::Subscription<Heartbeat>::SharedPtr hb_sub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  rclcpp::Subscription<fsd_msgs::msg::ConeMap>::SharedPtr map_sub_;
  rclcpp::Subscription<fsd_msgs::msg::VehicleCmd>::SharedPtr cmd_sub_;
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr check_timer_, keepalive_timer_;
  std::unique_ptr<fsd::HeartbeatEmitter> hb_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SafetySupervisorNode>());
  rclcpp::shutdown();
  return 0;
}
